File: JobApplicationsMicroservice/src/main/java/com/licenta/jobapplicationmicroservice/business/service/DatabaseService.java
package com.licenta.jobapplicationmicroservice.business.service;

import com.google.gson.Gson;
import com.licenta.jobapplicationmicroservice.business.interfaces.IDatabaseService;
import com.licenta.jobapplicationmicroservice.business.model.Company;
import com.licenta.jobapplicationmicroservice.business.model.Job;
import com.licenta.jobapplicationmicroservice.business.util.errorhandler.RestTemplateResponseErrorHandler;
import com.licenta.jobapplicationmicroservice.business.util.exception.ExceptionWithStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

@Service
public class DatabaseService implements IDatabaseService {

    private final RestTemplate restTemplate;
    private final String BASE_URL = String.format("http://%s:%s/api", System.getenv("DATABASE_MICROSERVICE_HOST"), System.getenv("DATABASE_MICROSERVICE_PORT"));
    private final String URL_FORMAT = "%s/%s/%s";

    @Autowired
    public DatabaseService() {
        restTemplate = new RestTemplateBuilder()
                .errorHandler(new RestTemplateResponseErrorHandler())
                .build();
    }

    @Override
    public void verifyUserExists(Long userId) {
        ResponseEntity<String> response = restTemplate.getForEntity(String.format(URL_FORMAT, BASE_URL, "users", userId), String.class);

        if(response.getStatusCode() != HttpStatus.OK)
            throw new ExceptionWithStatus(response.getBody(), response.getStatusCode());
    }

    @Override
    public Job getJob(Long jobId) {
        ResponseEntity<String> response = restTemplate.getForEntity(String.format(URL_FORMAT, BASE_URL, "jobs", jobId), String.class);

        if(response.getStatusCode() != HttpStatus.OK)
            throw new ExceptionWithStatus(response.getBody(), response.getStatusCode());

        Gson gson = new Gson();
        return gson.fromJson(response.getBody(), Job.class);
    }

    @Override
    public Company getCompany(Long companyId) {
        ResponseEntity<String> response = restTemplate.getForEntity(String.format(URL_FORMAT, BASE_URL, "companies", companyId), String.class);

        if(response.getStatusCode() != HttpStatus.OK)
            throw new ExceptionWithStatus(response.getBody(), response.getStatusCode());

        Gson gson = new Gson();
        return gson.fromJson(response.getBody(), Company.class);
    }
}

File: NewsfeedMicroservice/src/main/kotlin/com/licenta/newsfeedmicroservice/business/services/DatabaseService.kt
package com.licenta.newsfeedmicroservice.business.services

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.licenta.newsfeedmicroservice.business.interfaces.IDatabaseService
import com.licenta.newsfeedmicroservice.business.model.*
import com.licenta.newsfeedmicroservice.business.util.exception.ExceptionWithStatus
import khttp.get
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class DatabaseService: IDatabaseService {
    private val baseUrl = "http://${System.getenv("DATABASE_MICROSERVICE_HOST")}:${System.getenv("DATABASE_MICROSERVICE_PORT")}/api"

    override fun getCompanyIndustriesFollowedByUser(userId: Long): List<CompanyIndustry> {
        val response = get("${baseUrl}/users/${userId}/followed-company-industries")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<List<CompanyIndustry>>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }

    override fun getCityById(cityId: Int): City {
        val response = get("${baseUrl}/cities/${cityId}")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<City>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }

    override fun getCompanyById(companyId: Long): Company {
        val response = get("${baseUrl}/companies/${companyId}")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<Company>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }

    override fun getCountryById(countryId: Int): Country {
        val response = get("${baseUrl}/countries/${countryId}")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<Country>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }

    override fun getWorkTypeById(workTypeId: Int): WorkType {
        val response = get("${baseUrl}/work-types/${workTypeId}")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<WorkType>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }
}